### ISSUE TITLE
Add automatic bias computation in LinearMachine

### ISSUE THREAD

**Issue by youssef-emad** [shogun-toolbox/shogun#3033]
Adding Bias calculation to `train_machine`.
Replacing`cblas_daxpy` call by a `linalg` call or by using Eigen3.

**Comment by youssef-emad** [comment]
@karlnapf I have a question in the [train_machine()](http://shogun-toolbox.org/doc/en/latest/LinearRidgeRegression_8cpp_source.html#l00043) method.
what is the point of using a matrix (`kernel_matrix`)to hold the weights instead of using a vector?

**Comment by youssef-emad** [comment]
@karlnapf I have one question in [train_machine()](http://shogun-toolbox.org/doc/en/latest/LinearRidgeRegression_8cpp_source.html#l00043).
I don't understand the use of kernel_matrix instead of vector to hold weights.

**Comment by karlnapf** [comment]
What is the point of opening this exactly? :)

**Comment by karlnapf** [comment]
The point you asked why the kernel matrix is stored is simply: because the solution depends on a linear system involving the kernel matrix

**Comment by youssef-emad** [comment]
@karlnapf yeah the question about the matrix was silly so I erased it.
I opened it because I intend to make the desired changes so just to separate it from the main issue.

**Comment by karlnapf** [comment]
Don't worry about such questions ;)
Yes definitely this is a separate task. So I will leave this open until it is done

**Comment by karlnapf** [comment]
Renamed it, this should actually be done in the base class -- every linear algorithm should the option of pre-computing the data mean manually, and do it automatically by default

**Comment by youssef-emad** [comment]
@karlnapf but as I understand from the [docs](http://shogun-toolbox.org/doc/en/latest/classshogun_1_1CLinearMachine.html),the train_machine() method in CLinearMachine is just a virtual function re implemented in the subclasses.

**Comment by youssef-emad** [comment]
@karlnapf I edited the train_machine() method in LinearRidgeRegression class. check [here](https://github.com/youssef-emad/shogun/blob/29230ba37b496fe6179dd996891f7f4bdd7441cb/src/shogun/regression/LinearRidgeRegression.cpp#L45).
I also made a [notebook](https://github.com/youssef-emad/shogun_notebooks/blob/master/LinearRegressionNew.ipynb) with results using the same datasets.
I got rid of `lapack` calls and replaced them with `eigen`.
For the bias calculation, I set to the mean of error between the true target and the predication with bias equals zero.
e.g
```
true target is vector y
y = bias + w*feats
bias = mean( y - w*feats)
```
that's what I understood when you mentioned the data mean because I don't think you meant the mean of the labels because if I have a dataset with labels having a real intercept zero but only positive values the calculated intercept will be wrong ( e.g first example in notebook). Am I getting it right ?

**Comment by karlnapf** [comment]
sure it is a virtual function, but computing the data mean still should not happen in specialised classes, but in the base class. We can always add an additional method that calls a virtual method that is overloaded in subclasses...

**Comment by karlnapf** [comment]
BTW please don't post code with gists, send a PR. Much easier to review

**Comment by karlnapf** [comment]
In the notebook, why are the results not *exactly* the same? Can you check that the ridge has the same value?

**Comment by karlnapf** [comment]
What about larger datasets and dimensions? Timing might change then?

**Comment by karlnapf** [comment]
Ok had a look at the code. The bias addition should happen in the base class as said, and it should be possible to set the bias manually and disable that it is computed from data.
Can you send a PR with the changes? I have a few more comments, but don't want to put them in your branch

**Comment by youssef-emad** [comment]
@karlnapf The results are exactly the same for the second and the third datasets because they are fixed but the first is randomly generated each time so the results will change ( sklearn results also changes for the first dataset).
For the bias calculation,so I will make a function that computes the bias in the base class but it's not clear  to me how it will be called without modifying the train_machine in subclasses.

**Comment by karlnapf** [comment]
The sub-classes will not compute the bias as it currently has to be set manually.
Therefore, one way to do this would be to rename the train_machine methods (that dont compute bias) and add an additiona lmethod compute_bias() in the base class. Train_machine in the base class then (optionally) computes the bias and then calls the train_machine_no_bias of the specialsed classes.
Something like this, standard polymorhpic OOP

**Comment by youssef-emad** [comment]
@karlnapf I got it but I have a suggestion. Instead of renaming the train_machine in all subclasses can I override the train() method inherited from the CMachine class and add the optional compute_bias call to it.

**Comment by karlnapf** [comment]
Thats Indeed much better and actually the reason why we have this structure in the base class

**Comment by youssef-emad** [comment]
I think this issue can be closed :) , but I have one question for my understanding.
Why train_machine is implemented in a closed form solution instead of using gradient descent or stochastic gradient descent. From my basic knowledge I think computing X transpose and multiplying X by X transpose may be very expensive operations specially for large number of features like 100,000 feature or something like that.

**Comment by karlnapf** [comment]
Yes, the matrix is DxD where D is number of features.
We could implement a stochastic version of this, but even better would be to use an existing library to do so.....Vowpal wabbit for example
